Easy Stacks one-pager

muvaf opened the pull request:
<!--
Thank you for helping to improve Crossplane!

We strongly recommend you look through our contributor guide at https://git.io/fj2m9
if this is your first time opening a Crossplane pull request. You can find us in
https://slack.crossplane.io/messages/dev if you need any help contributing.
-->

### Description of your changes

Introducing the concept of _easy stacks_ which allow a stack to be built with a set of default managed resources and resource classes and then deployed to the cluster easily in order to provide a ready-to-use environment to the user easily.

The one-pager lays down the general principles to adhere when developing such stacks. #1096 tracks what the sets of YAMLs will be in the initial implementations of easy stacks. Depending on timeline, a generic controller will be implemented following the implementation of the initial stacks.

Fixes https://github.com/crossplaneio/crossplane/issues/1095
<!--
Briefly describe what this pull request does. Be sure to direct your reviewers'
attention to anything that needs special consideration.

We love pull requests that resolve an open Crossplane issue. If yours does, you
can uncomment the below line to indicate which issue your PR fixes, for example
"Fixes #500":

Fixes #
-->

### How has this code been tested?
<!--
Before reviewers can be confident in the correctness of a pull request,
it needs to tested and shown to be correct. In this section, briefly
describe the testing that has already been done or which is planned.
-->
Manual tests around whether the statements would actually make sense when implemented.
### Checklist
<!--
Please run through the below readiness checklist. The first two items are
relevant to every Crossplane pull request.
-->
I have:
- [ ] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Ensured this PR contains a neat, self documenting set of commits.
- [ ] Updated any relevant [documentation] and [examples].
- [ ] Reported all new error conditions into the log or as an event, as
  appropriate.

For more about what we believe makes a pull request complete, see our
[definition of done].

[documentation]: https://github.com/crossplaneio/crossplane/tree/master/docs
[examples]: https://github.com/crossplaneio/crossplane/tree/master/cluster/examples
[definition of done]: https://github.com/crossplaneio/crossplane/tree/master/design/one-pager-definition-of-done.md

prasek commented on design/one-pager-easy-stacks.md:
```diff
@@ -0,0 +1,294 @@
+kind: EasyAWS
+metadata:
+  name: project-future
+  namespace: dev
```
per comment above -- an environment (dev/test, stage, prod) generally maps to crossplane instance, so could this be cluster-scoped? If not perhaps we should recommend it go into the crossplane-system namespace [like we do for provider secret today](https://github.com/crossplaneio/crossplane/blob/master/docs/cloud-providers/aws/aws-provider.md#optional-setup-aws-provider-manually).

muvaf commented on design/one-pager-easy-stacks.md:
```diff
@@ -0,0 +1,294 @@
+As sky is the limit when it comes to stacks, we need to put down some general
+rules to be adhered by initial configuration stacks.
+
+* CRDs of the resources that will be deployed by the stack are declared as
```
Yes, like this example https://github.com/crossplaneio/crossplane/blob/master/design/design-doc-stacks.md#example-appyaml

displague replied:
The deployed resources of a Stack are not contained in the `dependsOn` block, they do not exist in the `app.yaml` file.  The StackManager walks the CRD files in the Stack OCI, installs them all, and creates a Stack record with a `customresourcedefinitions` list (as in `stack/stack-gcp`):

```yaml
 customresourcedefinitions:
  - apiVersion: cache.gcp.crossplane.io/v1beta1
    kind: CloudMemorystoreInstanceClass
  - apiVersion: cache.gcp.crossplane.io/v1beta1
    kind: CloudMemorystoreInstance
  - apiVersion: compute.gcp.crossplane.io/v1alpha3
    kind: GKEClusterClass
  - apiVersion: compute.gcp.crossplane.io/v1alpha3
    kind: GKECluster
  - apiVersion: compute.gcp.crossplane.io/v1alpha3
    kind: GlobalAddress
  - apiVersion: compute.gcp.crossplane.io/v1alpha3
    kind: Network
  - apiVersion: compute.gcp.crossplane.io/v1alpha3
    kind: Subnetwork
  - apiVersion: database.gcp.crossplane.io/v1beta1
    kind: CloudSQLInstanceClass
  - apiVersion: database.gcp.crossplane.io/v1beta1
    kind: CloudSQLInstance
  - apiVersion: gcp.crossplane.io/v1alpha3
    kind: Provider
  - apiVersion: servicenetworking.gcp.crossplane.io/v1alpha3
    kind: Connection
  - apiVersion: storage.gcp.crossplane.io/v1alpha3
    kind: BucketClass
  - apiVersion: storage.gcp.crossplane.io/v1alpha3
    kind: Bucket
  dependsOn:
  - crd: '*.cache.crossplane.io/v1alpha1'
  - crd: '*.compute.crossplane.io/v1alpha1'
  - crd: '*.database.crossplane.io/v1alpha1'
  - crd: '*.core.crossplane.io/v1alpha1'
  - crd: '*.storage.crossplane.io/v1alpha1'
  - crd: '*.workload.crossplane.io/v1alpha1'
```

muvaf replied:
>  they do not exist in the app.yaml file

@displague I guess the design doc is not updated with the different implementation, right?

muvaf commented on design/one-pager-easy-stacks.md:
```diff
@@ -0,0 +1,294 @@
+    the stack uses and expose the high level variations of the parameters through
+    `spec` instead of having different custom resource definitions for each class
+    of configuration.
+* A minimal high level set of configurations are exposed to the user through
```
> if this will include passing the provider secret content via environment config CR

Not sure about passing down the secret _content_ via CR.

I think taking a `Secret` name and namespace in a field like `spec.credentialsSecretRef` is the easiest option. This way the client that will put the credentials doesn't have to know how to construct a specific `Provider` CR. Client will;
1. Create a secret.
2. Create a `EasyAWS` CR with `spec.credentialsSecretRef` that points to the created secret.

> requiredResources:
> resource:

I couldn't find a reference to these properties of `app.yaml`. Could you point me to a resource that explains how they work?

> Note that maxAllowed would limit the number of environment config CRs created, at least for now as a simplifying assumption, unless it would make sense to allow multiple env. config CRs for a given config stack in the first release.

Limiting the number of CRs to 1 would indeed simplify the initial implementation. But I'm not aware of a way to tell api-server not to accept creation call of specific CR after some point.

hasheddan replied:
> But I'm not aware of a way to tell api-server not to accept creation call of specific CR after some point.

As of Kubernetes 1.15 you can use a `ResourceQuota` object to [restrict the count](https://kubernetes.io/docs/concepts/policy/resource-quotas/#object-count-quota) of custom resource instances in a namespace. Maybe we could create `ResourceQuota` objects on stack installation based on the `maxAllowed` fields!

suskin replied:
> I think taking a Secret name and namespace in a field like spec.credentialsSecretRef is the easiest option.

I agree with this. Creating a secret separately and then putting a reference in the spec field of the single CR seems like it would be more manageable than requiring multiple resources (and coordinating that).

negz replied:
> I think taking a Secret name and namespace in a field like spec.credentialsSecretRef is the easiest option.

Throwing my support behind this too - Secrets are the right place for credentials. Let's not include sensitive credentials inline in our custom resources; it shouldn't be too onerous to create a (provider credentials) `Secret` and an `EasyAWS` that references it in one fell swoop.

> Limiting the number of CRs to 1 would indeed simplify the initial implementation.

I can't recall ever coming across an example of a singleton resource in Kubernetes (e.g. only one instance of `kind: Foo` is allowed per namespace). I suspect that it would not be significantly more work to cleanly handle multiple instances of each easy stack custom resource being created in one namespace, and also suspect that the user experience would be fragile/surprising/weird if we _did not_ support this.

muvaf replied:
From comments and our chat yesterday, the direction we're taking is expose a field to accept a secret ref and support multiple CR instances.

muvaf commented on design/one-pager-easy-stacks.md:
```diff
@@ -0,0 +1,294 @@
+* **TBD** Controller should update the resources continuously and treat the given CR as
+  the source of truth even though user manually changes the resources that are
+  deployed by the stack.
+* **TBD** CRD of the configuration stack should be namespaced so that different
```
How I imagined is like the following; the users and UI clients can create as many config CRs as they want. To prevent the name collisions of deployed cluster-scoped resources, we'd prefix the names with `namespace-name-` of the specific config CR. This way people can have many configuration environments independently in one Crossplane instance.

This thinking treats the creator of the config CR as app user rather than infra owner role that  who has cluster-wide access. Technically, implementing the CRDs namespaced or cluster-scoped requires same amount of effort. So, if the creator is assumed to be the infra owner and we don't expect app developers to create config CRs, we can go with cluster-scoped.

hasheddan replied:
+1 for cluster-scoped. I think that these stacks could be used for base-level infrastructure across multiple teams, so it seems likely they will likely be deployed by infra-owners in most cases.

negz replied:
Dropping a note here that if these easy stack custom resources are namespaced we won't be able to rely on garbage collection to cleanup the cluster scoped managed resources and classes they create, thus introducing a touch more complexity to handle cleanup. I'm thinking we should assume installing an easy stack is the domain of the infrastructure operator (and thus they're cluster scoped) unless we think it's really compelling to allow app owners to install them.

muvaf replied:
Cluster-scoped sounds good.

hasheddan commented on design/one-pager-easy-stacks.md:
```diff
@@ -0,0 +1,294 @@
+* Apply; continuously reproduce the resources from YAMLs and apply it.
+* Deletion: should be done via label selector when the CR is deleted.
+
+In the `Apply` phase, there will be some processing phasse before calling `Create`
```
```suggestion
In the `Apply` phase, there will be some processing phase before calling `Create`
```

hasheddan commented on design/one-pager-easy-stacks.md:
```diff
@@ -0,0 +1,294 @@
+Today, it's possible to do all this by manually configuring each resource YAML
+files and run `kubectl apply -f <directory of YAMLs>`. However, there are two areas
+that we can improve during this process:
+* The preparation phase of these YAMLs requires to be very careful and changing
```
```suggestion
* The preparation phase of these YAMLs requires a user to be very careful and changing
```

suskin commented on design/one-pager-easy-stacks.md:
```diff
@@ -0,0 +1,294 @@
+
+## Goals
+
+The main goal of this design is to make it easier for people to write sets of
```
> main goal of this design

I would also try to mention the goal for the end user, which is to make it easier to set up a working environment.

muvaf replied:
Changed that section to be a bit more generic.

suskin replied:
I agree with the point @negz is making about it being helpful if we are more specific about which end-user persona we're targeting. I'm making a note of this because it appears that the goal is not yet written in that way.

suskin commented on design/one-pager-easy-stacks.md:
```diff
@@ -0,0 +1,294 @@
+
+* A base boilerplate package where creating a new configuration set that has a
+  controller is as easy as changing the YAML files.
+* A way to declare pre-requisites that needs to be satisfied by the cluster
```
What is the benefit of including this in the design? I would not expect a formal mechanism to be needed at this point, since people can just install things in the right order according to some documentation. Is this referring to listing the crds it needs to be able to CRUD?

muvaf replied:
I was under the misconception that the stack `app.yaml` actually provided that ability. But I see that's not the case https://github.com/crossplaneio/crossplane/pull/1110#discussion_r352728755

So, I'll just remove that and expect people to have installed all the CRDs needed.

suskin commented on design/one-pager-easy-stacks.md:
```diff
@@ -0,0 +1,294 @@
+  controller is as easy as changing the YAML files.
+* A way to declare pre-requisites that needs to be satisfied by the cluster
+  for the configuration set to work.
+* A way to change a high level configuration by changing 1 value that is then
```
Grammar/style nitpick: normally the numbers one through ten would be written out, so `changing one value` in this case

suskin commented on design/one-pager-easy-stacks.md:
```diff
@@ -0,0 +1,294 @@
+* It should be possible to have different sets of configurations in one stack
+  and user should be able to choose from them via the CR that is an instance
+  of stack's CRD.
+  * Note that it's highly preferred to have 1 custom resource definition that
```
> Note that it's highly preferred to have 1 custom resource definition

What is the reason for this preference?

suskin replied:
Also, I am thinking that we may want to require that it is a single CRD at this time, because it seems like in order to support multiple we would need a lot of the configuration that we are designing as part of the template stacks work.

muvaf replied:
This has to do with how we name the deployed resources. Basically, when you have only one CRD, you can just prefix the resources with the CR instance name. The api-server would reject creation of CR instances with the same name, so, name as prefix is enough. When you have multiple CRDs, name is not unique anymore. So, you either use `uid` or some other method for making deployed resources exclusive to one CR that intended to deploy them.

However, it's also possible that you can have multiple CRDs that deploy completely different kind sets of resources, you'd not need to worry about the name collision in that case. So, it's not that you should never do it but it's just highly recommended to go with one custom resource definition and build another stack for that extra CRD that you want.

suskin commented on design/one-pager-easy-stacks.md:
```diff
@@ -0,0 +1,294 @@
+* A minimal high level set of configurations are exposed to the user through
+  the stack's CR, such as `Provider` reference to refer in the deployed resource
+  or `region`.
+* Controller should signal readiness of all resources in CR's status.
```
What is `readiness`? How does the controller know they are ready?

negz replied:
Do we get much value from modelling the fact that the entire configuration set is ready?

Resource classes have no concept of readiness - if they exist they're ready - so I imagine this would be communicating that the various supporting managed resources (VPCs, etc) were ready to use. Will anything block on this readiness (e.g. you can't create a `KubernetesApplication` until the `EasyAWS` is ready), or would it just be a signal to the infrastructure operator that everything came up okay?

muvaf replied:
> What is readiness? How does the controller know they are ready?

Its details are in `Reconciliation` section.

> Do we get much value from modelling the fact that the entire configuration set is ready?

Thinking about it again, not a lot really. It'd help the infra-owner to know everything is ready but that'd be just informational. I'll remove that part.

muvaf replied:
Oh, actually I think if we remove that there will be no `status`, meaning no way for anyone to know the stuff is actually working. So, I think we'd want to have at least the `conditions` part of the status with reconciliation status.

suskin replied:
It does seem valuable to have a way to check the status of everything in a single place. However, I am not sure we need to make it a requirement for resource packs. Since the main goal is the focus on applying a set of configuration best practices without writing your own, that goal seems somewhat orthogonal to the goal of seeing the statuses in one place.

Put another way, I think it is an improvement even without the status aggregation, so we don't necessarily need to include the status aggregation in the first version. For example, it is possible that we add status/readiness aggregation in some other way, with a UI or with some `trace`-like functionality.

muvaf replied:
@suskin I have added `ConditionedStatus` to status of the initial gcp implementation. We'll have `Synced` type condition to be set by the generic controller.

suskin commented on design/one-pager-easy-stacks.md:
```diff
@@ -0,0 +1,294 @@
+* Stack type will be `ClusterStack` as most of the resources it deploys are
+  cluster-scoped.
+
+Name of the concept is **TBD**; _EasyStack_, _ConfigurationStack_,
```
"Resource Pack" is another candidate, and my personal preference : ). I am also liking the "Configuration Stack" option though (or "Configuration Pack").

negz replied:
https://www.slideshare.net/SarahZelechoski/how-up-running-culture-is-devaluing-engineering-work

I watched the above presentation recently and it resonated with me. It cautions against branding tools that partially abstract _really complicated stuff_ (like deploying and managing cloud infrastructure) as making those things "easy" - see the deck for more context. Therefore I'd prefer a name that either puts no particular value on what the thing does like `ResourcePack` (this is a pack of resources), or something that emphasises a little more specifically the value they bring. I think they intend to quickly provide infrastructure operators a basic configuration foundation on which to build, so something like `QuickConfig`, or `BaseConfig` could make sense.

negz replied:
Are there any emergent patterns around stack names and the custom resources they install?

There's a few _classes of stack_, e.g. "infrastructure stack", "application stack", "template stack". Then there are specific instances of each class of stack, like `stack-gcp`, `stack-azure`, and `stack-wordpress`. We don't have data yet but I suspect many app stacks will be named `stack-foo` and define exactly one custom resource of `kind: Foo`.

So, perhaps:

* This new class of stack is an "infrastructure config stack".
* Each instances of an infrastructure config stack is named like `stack-aws-config-upbound`, or `stack-aws-config-acme` ("this the Upbound variant of the AWS infrastructure configuration stack").
* They each define one custom resource something like `kind: AWSConfigUpbound`.

🤷‍♂

muvaf replied:
@suskin I think the word _stack_ should be in it somewhere. I am leaning towards `Configuration Stacks`.

@negz I feel like `Config` and `Upbound` make the stack names too long. How about? `kind: AWSMinimal group: aws.configurationstacks.upbound.io`

suskin replied:
> It cautions against branding . . . as making those things "easy"

I agree with @negz wholeheartedly on this. I have a strong dislike of the name "Easy", and the presentation Nic linked explains the reasons better than I was able to articulate them.

> or something that emphasises a little more specifically the value they bring

I see the general name (e.g. Resource Pack, Configuration Stack) as the name for the style of stack, and I see something like the proposed "Base Config" as a more specific term to describe the value that a specific pack/stack is providing. I see them as complementary, and not exclusive. I see it as a corollary to how a `GKECluster` resource is a GKECluster, and is a managed resource, and is a resource.

> I am leaning towards `Configuration Stacks`

I slightly prefer "pack" over "stack" because "stack" just happens to be the format that we're using to bundle it, whereas "pack" describes what it is conceptually, without tying it to a particular format. That makes it more accessible for people who are not yet familiar with our idioms and terms. My **favorite** is "resource pack", but I also like "configuration pack". I slightly prefer "resource" because it is more literal about what is in the pack (a bunch of resources), whereas "configuration" could be interpreted as referring to other types of configuration.

suskin commented on design/one-pager-easy-stacks.md:
```diff
@@ -0,0 +1,294 @@
+### Metadata of The Stack
+
+[Dependency list] has to be filled with each kind of the resource that the stack
+deploys, i.e. f the stack only deploys `RDSInstanceClass` resource, it should
```
`i.e. f` seems like it should be `i.e. if`.

suskin commented on design/one-pager-easy-stacks.md:
```diff
@@ -0,0 +1,294 @@
+it's good to separate the folders cleanly so that it's easier to fork & change.
+
+There are mainly 2 types of resources that configuration stacks will usually
+need to deploy; resource classes and managed resources. Under `resources` folder
```
Grammar nitpick: I would normally use a colon when giving an example or a list of examples, so `deploy: `. Semicolons (`;`) are for [connecting related clauses](https://en.wikipedia.org/wiki/Semicolon#Usage); often a semicolon can be used where a period would also work.

suskin commented on design/one-pager-easy-stacks.md:
```diff
@@ -0,0 +1,294 @@
+need to deploy; resource classes and managed resources. Under `resources` folder
+we can have 2 separate folders for each of those resource types; `classes` and
+`services`. Under these folders, resources should be in directories recursively
+named after their group and kind. An example folder structure:
```
Maybe I am misunderstanding, but it doesn't seem like the example folder structure matches "`directories recursively named after their group and kind`". It seems more like the folder is the group, and the filename is `{{kind}}.yaml`. Am I reading the example correctly?

muvaf replied:
I am not 100% sure about this one actually. Going as deep as having a folder for each `kind` looked too verbose to me. What do you think about having `resources/classes/<groupname>/<resource name>.yaml` ? For resources of different kinds with same names, the author probably would want to make the file name something like `<kind>-<object name>.yaml`

suskin replied:
I don't think we need to enforce any particular structure. I would provide some examples of what we feel are best practices for logical groupings, but keep the tooling unopinionated and leave it up to the author to decide what works well for them. If it's supported by the tool they're using underneath (in this case, kustomize), then we should support it. The grouping that we're using in the example structure looks pretty reasonable to me.

suskin commented on design/one-pager-easy-stacks.md:
```diff
@@ -0,0 +1,294 @@
+
+Example above is a suggestion for stacks with one tier of configuration.
+There could also be cases where one configuration stack has the YAML files for
+different tiers and a `spec` allows to deploy one of them, in that case it'd make
```
> a `spec` allows to deploy one of them

Hmm, I'm wondering how that would be implemented. But, I'm not sure we need to figure that out before merging this design.

muvaf replied:
It can filter the resources with labels while reading the YAML files.

suskin commented on design/one-pager-easy-stacks.md:
```diff
@@ -0,0 +1,294 @@
+### Reconciliation
+
+There are mainly 3 responsibility of the controller:
+* When all deployed resources are ready, `status.state` should be `ready`.
```
I like this

suskin commented on design/one-pager-easy-stacks.md:
```diff
@@ -0,0 +1,294 @@
+  * For other possible resources, the developer chooses what to look for as
+    readiness.
+* Apply; continuously reproduce the resources from YAMLs and apply it.
+* Deletion: should be done via label selector when the CR is deleted.
```
Are we able to use garbage collection? I like the label idea, but I want to check.

muvaf replied:
It's not possible if the configuration stack CRD is namespace-scoped. But the direction is to go with making it cluster-scoped, so I will switch that to owner references, which is easier to implement and maintain.

suskin commented on design/one-pager-easy-stacks.md:
```diff
@@ -0,0 +1,294 @@
+for the resources:
+
+1. Names of the resources should be converted to the following format
+  `<CR namespace>-<CR name>-<Name given in YAML>` to prevent name collisions
```
Is there a name length limit, and could we run into problems? Another option is to include the CR UID in the name, though it's a little less human-friendly.

muvaf replied:
I believe 63 is the character limit. Since it's now cluster-scoped, we'll only use the CR's name and I think `uid` is longer than most of the CR names. So, it should be safer to go with name.  But, in general, yes if we hapen to have very long CR name and resource name, we'll run into a problem.

suskin commented on design/one-pager-easy-stacks.md:
```diff
@@ -0,0 +1,294 @@
+  So, the name that appears on YAML files are valid for referencing each other
+  while building the stack but in actual operation, do not expect exact names
+  to appear.
+  * This will require cross-resource references to be scanned and replaced with
```
Hmm, this feels similar to having a templated name; what are your thoughts on using explicit variables to represent the values that will be injected into the yamls? That way, the yamls are closer to what will be applied, which people will like for predictability.

negz replied:
I'd like to understand a little better how the controller would scan and replace these cross resource references. If the controller is operating on raw YAML (e.g. `*unstructured.Unstructured`) I suspect it would be pretty hard to identify them and determine what they should be referencing.

negz commented on design/one-pager-easy-stacks.md:
```diff
@@ -0,0 +1,294 @@
+  providerRef:
+    name: aws-provider
+status:
+  state: ready
```
Most Crossplane resources today use a conditioned status to represent readiness. Should we do so here?

muvaf replied:
Yes, that makes sense. Per discussion on https://github.com/crossplaneio/crossplane/pull/1110#discussion_r353461481 I think representing a condition for reconciliation status could be enough.

negz commented on design/one-pager-easy-stacks.md:
```diff
@@ -0,0 +1,294 @@
+    considered to be ready.
+  * For other possible resources, the developer chooses what to look for as
+    readiness.
+* Apply; continuously reproduce the resources from YAMLs and apply it.
```
Do you expect there will be one generic controller loading this YAML as `*unstructured.Unstructured`, or a different controller for each stack variant that knows the actual kinds it supports at the code level?

It doesn't seem like the controller will really care much about the shape of the resources with two exceptions - it will want to add metadata (e.g. set labels) when creating resources, and will need to know how each resource reflects it readiness.

muvaf replied:
> Do you expect there will be one generic controller loading this YAML as *unstructured.Unstructured, or a different controller for each stack variant that knows the actual kinds it supports at the code level?

Became unrelated as we started to use `kustomize`. But in patchers, we do make the operations on `*unstructured.Unstructured` type.

> will need to know how each resource reflects it readiness.

No readiness anymore. If the reconciliation is completed with no errors, it means the resources are deployed and we tell that to user via `Synced` condition.

muvaf commented on design/one-pager-easy-stacks.md:
```diff
@@ -0,0 +1,294 @@
+## Goals
+
+The main goal of this design is to make it easier for people to write sets of
+configurations and be able to manage those configuration sets on a higher level.
```
You deploy a set of configurations, which is a bunch of YAML files. Then manage them through the CR. _Managing_ in this context usually means controlling their lifecycle via create/update/delete methods. Update could cover a surface as minimal as changing the credential secret or as big as changing the region on each and every resource.

muvaf commented on design/one-pager-easy-stacks.md:
```diff
@@ -0,0 +1,294 @@
+that we can improve during this process:
+* The preparation phase of these YAMLs requires to be very careful and changing
+  something usually means touching a few different files. While this may not seem
+  like a big hurdle, changing some high level parameters of that environment is
```
Usually, it's a parameter that affects more than one resource. For example, region parameter that exist in many resources could be controlled on CR, controller would go ahead and make the changes on each resource. But it doesn't have to be that wide-affecting, it's basically up to the developer to decide what is high level/low level parameter.

negz commented on design/one-pager-configuration-stacks.md:
```diff
@@ -70,12 +70,10 @@ configurations and be able to manage those configuration sets on a higher level.
-* A way to change a high level configuration by changing 1 value that is then
-  translated to the YAMLs that are already deployed.
+* A way for infrastructure owners to write their configuration sets with
+  reconciliation easier than writing a controller from scratch.
```
It feels like we're really close to having something that allows infrastructure owners to write their configuration sets without writing _any_ controller. Should that be a goal?

negz commented on design/one-pager-configuration-stacks.md:
```diff
@@ -254,30 +244,35 @@ sense for developer to have a folder for each tier under `resources` folder.
-  since most of the resources stack deploys are cluster-scoped.
-  So, the name that appears on YAML files are valid for referencing each other
+1. Names of the resources should be converted to the processing phase following
+  format `<CR name>-<Name given in YAML>` to prevent name collisions.
```
Is there something we can do to make it more obvious that this transformation will happen? I worry that it could be surprising.

It seems like the stack author will write something like:

```yaml
providerRef:
  name: cool-provider
```

But the resource we actually create will be modified to:

```yaml
providerRef:
  name: mycoolcustomresource-cool-provider
```

negz commented on design/one-pager-configuration-stacks.md:
```diff
@@ -288,6 +283,10 @@ for the resources:
   * Stacks do a better job for declaring CRD dependencies and UI annotation metadata.
+* Using kustomize or Go templates could save us from adding a Go function for each
+  kind we add even though this'd limit the future addition abilities it'd could
+  be preferrable.
```
I have an inkling that using kustomize could be elegant here. Something like:

* User provides the base YAML for a bunch of resources.
* Our controller generates an overlay that sets resource names,  labels, and cross-resource reference.
* Our controller applies the generated overlay to the base layer and applies it to the API server.

This could allow folks to fork and define their own minimal/easy stacks simply by writing YAML (and stack scaffolding metadata). It seems like there's a bunch of potential overlap with template stacks here, but I don't think that needs to block this effort. It could be fine to write a purpose-built controller that uses Kustomize to do this specific task, then assess replacing it with template stacks if and when they're ready.

muvaf replied:
My initial intent was to just use Kustomize for all transformations and let the controller be generic. However, I had gotten stuck while trying to do it, especially the part where you declare a reference to be filled by X kind that exists in the pack with that name.

I went through a few examples, dug deeper and finally found that there is a support for CRDs to declare custom reference fields. https://github.com/kubernetes-sigs/kustomize/blob/master/examples/transformerconfigs/crd/README.md

I have converted the minimal GCP YAML files into that structure and it works very well. We move the burden of the reference handling from the controller to the `kustomizeconfig.yaml` file. Note that we still need to add some lines to that file for each _kind_ that we add. But I guess it's better than adding a Go function.

In fact, we can have a giant `kustomizeconfig.yaml` file in `stack-gcp` that declares the dependencies for each type and fetch it, so, author will not worry about the reference bindings at all.

The controller will use the `kustomization.yaml` file that has `apiVersion: kustomize.config.k8s.io/v1beta1 kind: Kustomization` metadata. This introduces a dependency on kustomize but as far as I see, it's already pretty much the standard in the ecosystem for configuration management and templating. It's even in `kubectl`. So, it should be fine to depend on it IMO.

@prasek what do you think? The controller will be generic but the developer experience will change from dumping all yamls to dumping a kustomize directory and we'll depend on kustomize. Keep in mind that this is pretty much an implementation detail, nothing will change for clients that create the CR, i.e. Upbound backend.

muvaf replied:
@negz There is one caveat: the custom transformer config identifies resources with only `name` and `kind`, it doesn't include `apiVersion` during its search. So, the resources that share the same `kind` but has different `apiVersion`, i.e. multicloud conf stacks, will be treated in the same way.

This would cause problem only if say the stack has one AWS Provider and GCP Provider that share the same name. For other resources, most of the time `kind` is different, like VPC is named `VPC` in AWS but `Network` in GCP. So, it should be fine except for `Provider` kinds which share the same kind across infra stacks.

Opened https://github.com/kubernetes-sigs/kustomize/issues/1885 to track this.

prasek replied:
@muvaf using a kustomize directory sounds good and is aligned with the template stacks direction towards pluggable templating/config engines including kustomize support, so this direction makes sense.

prasek commented on design/one-pager-configuration-stacks.md:
```diff
@@ -0,0 +1,282 @@
+kind: MinimalAWS
+metadata:
+  name: small-infra
+spec:
```
We'll also want to include spec attributes like region, projectId, and likely allow overriding defaults for things like CIDR / IP ranges for nodes, pods, services, etc. If we're using kustomize then this would likely generate a patch.yaml similar to what @hasheddan put together [here](https://github.com/crossplaneio/crosscd/blob/master/classpacks/workflow/argo/infra/gcp/patch.yaml).

muvaf replied:
The `patch.yaml` approach requires the types to be known by the controller and iteration over each and every one of them, which is not ideal IMO since the developer will have to implement more complex functions.

I've experimented with [transformer configs of Kustomize for variant referencing](https://github.com/kubernetes-sigs/kustomize/blob/master/examples/transformerconfigs/crd/README.md#adding-a-custom-resource) and I think it works for us. An example implementation is here https://github.com/muvaf/minimal-gcp/pull/2 . As you'll see, we define the region as `$(REGION)` in the YAML files and then in the controller code, only [define the path in the CR](https://github.com/muvaf/minimal-gcp/pull/2/files#diff-1fa7bf35a53cd97938369f8733221602R70) that the value should be taken.

prasek commented on design/one-pager-configuration-stacks.md:
```diff
@@ -0,0 +1,282 @@
+kind: MinimalAWS
+metadata:
+  name: small-infra
+spec:
```
We'll want to support creating resource classes marked as default, but if more than one MinimalAWS CRs are created, there should be a way to disable the default label being applied so there is only 1 set of default classes.

muvaf replied:
I have added a section about this. I suggest we have a field named `keepDefaultingLabels`(name suggestions are welcome) and if user intends to keep the defaulting label in resource classes, they have to make this field `true`. Otherwise, the controller will remove the defaulting labels whenever it sees one. So, by default, if several users deploy empty CRs they won't have any conflict. If they want the default, they'd need to explicitly say so.

prasek commented on design/one-pager-configuration-stacks.md:
```diff
@@ -0,0 +1,282 @@
+    kind: MinimalAWS
+    name: small-infra
+    uid: 1ca16960-973b-4d58-a6fa-5696638ec631
+  labels:
```
Will the `MinimalAWS` CR labels be copied down to the templated resources? Suggest adding an example of this.

muvaf replied:
Yes, I think that's a good UX improvement for users who use selectors in claims.

muvaf commented:
@negz https://github.com/muvaf/configuration-stacks/pull/1 is the initial implementation of the generic reconciler and https://github.com/muvaf/minimal-gcp/pull/1 is the first integration. It's unpolished, but I'd love your general feedback about the direction in generic reconciler.

@prasek I have added your comments as TODOs in the implementation PRs. I'll update the one-pager as soon as I can.

@suskin @prasek @negz I guess I agree with @suskin on the name `Resource Packs`. What do you guys think? Note that this will be visible to users both in the implementation as well as documentation.

muvaf commented on design/one-pager-easy-stacks.md:
```diff
@@ -0,0 +1,294 @@
+It is important that the design puts forward:
+
+* A base boilerplate package where creating a new configuration set that has a
+  controller is as easy as changing the YAML files.
```
Not exactly _forking_. The developer has to build the stack from ground up since they will implement their own `CRD`. I mean they can still fork the example stacks but then they'd have to use the same `CRD`.

Currently, we provide a generic reconciler that they can fetch with custom modification functions that applies their custom logics in different steps. How does the following statement sound?

```
A base tooling for infrastructure operators who would like to package a set of resources as a stack and provide that to their users with minimal Go code.
```

muvaf commented:
I have changed the name to `resource packs`. The PR https://github.com/muvaf/minimal-gcp/pull/2 is our first resource pack for minimal GCP. It has region and reference to cred secret as spec. Tested and it works as expected.

I think the major new thing is how we bind a field from the CR to the child resources that affects the developer experience. Other than that I will put comments in the code and write tests for the generic reconciler.

@prasek @negz @suskin IMO, both generic reconciler and minimal GCP looks like complete feature-wise for an initial implementation. Please let me know if you have suggestions that affect the user experience directly, otherwise I'd like to deliver that implementation as initial step and iterate as we go on.

negz commented on design/one-pager-configuration-stacks.md:
```diff
@@ -105,53 +106,77 @@ rules to be adhered by initial configuration stacks.
 * Stack type will be `ClusterStack` as most of the resources it deploys are
   cluster-scoped.
+* CRD of the resource pack should have a boolean spec field called
+  `keepDefaultingLabels` which indicates what happens to the resource classes that
```
Should this be annotations?

muvaf replied:
Yes, updated it.

negz commented on design/one-pager-configuration-stacks.md:
```diff
@@ -256,16 +292,130 @@ like the following:
+
+### Custom Patchers
+
+The resource pack reconciler that has two types of patcher functions where
```
I think the "that" on this line is redundant.

muvaf commented:
@negz I have updated this one-pager with the latest thinking. I think we're ready to merge this. The implementations are completed feature-wise, so I'd appreciate your review on:
* https://github.com/muvaf/crossplane-resourcepacks/pull/1 Generic reconciler
* https://github.com/muvaf/minimal-gcp/pull/2 Minimal GCP stack

After the merge I think we'll need to move the repositories to `crossplaneio` org.

As we synced with @suskin and @prasek, at some point, we're going to provide easy stacks via template stacks when that work lands. I added the code generation approach under the future considerations because it seems we'll not go in that direction. Though IMO, it'd have been really cool to have the `resources` folder and a YAML with descriptions about CRD and patchers and then just do `crossplane stack build . -t minimal-gcp` && `crossplane push minimal-gcp` with no Go code seen by the user.

negz commented on design/one-pager-resource-packs.md:
````diff
@@ -0,0 +1,508 @@
+1. Install GCP stack and create a secret that contains GCP credentials.
+2. Install a GCP resource pack via:
+```yaml
+# exact apiVersion is TBD.
````
Is this actually still TBD?
